---
title: "01-read_filter_normalize_scRNAseq_ALCL"
author: "Putri Ramadani"
date: "2025-05-28"
output: html_document
---

This is documentation for the script `scRNAseq_ALCL_Thymus_RLN_Adult_Ped-merged.Rmd`, which is part of the analysis pipeline for single-cell RNA sequencing data from ALCL patients. 
The script performs data import, filtering, and normalization.
The analysis pipeline is designed to handle multiple samples, perform quality control, and visualize the results through UMAP plots. The final output is a clustered Seurat object that can be used for further analysis and interpretation of the scRNAseq data.
Sample patients:
ALCL001
ALCL002
ALCL003
ALCL004
ALCL005
ALCL006
ALCL007
ALCL008
UHNALCL01
UHNALCL02
UHNALCL03
UHNALCL04

Control samples:
CLC02916 -> Reactive Lymph Node
CLC02917
CLC03483
CLC03510 (removed later due to low cell numbers)
CLC04353
EX01176 -> Thymus 
EX01177 (missing in the cluster, removed from metadata)
EX01178
EX01179
EX01180

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
This knitr chunk is used to set global options for the R Markdown document, such as whether to show code output or not.
echo = TRUE means that the code will be shown in the output document, while echo = FALSE would hide the code.

# 1. Load Libraries & Set Up
```{r}
library(Seurat)
library(pheatmap)
library(ggrepel)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readxl)
library(SoupX)
library(DropletUtils)

set.seed(1234)
setwd('~/scRNAseq_ALCL_Thymus_RLN')
lib.info <- read_xlsx('metadata/lib_ALCL_Thymus_RLN.xlsx')

dir.create("data", recursive = TRUE, showWarnings = FALSE)
dir.create("figures", recursive = TRUE, showWarnings = FALSE)
```


# 2. SoupX Decontamination + Seurat Object Creation
```{r}
sc.obj <- NULL
out.dir <- "data/SoupX_outputs"
dir.create(out.dir, recursive = TRUE, showWarnings = FALSE)

for (i in 1:nrow(lib.info)) {
  i.df <- lib.info[i,]
  sample_id <- i.df$Sample
  cat("Running SoupX on:", sample_id, "\n")

  filt.dir <- i.df$AnalysisDir
  raw.dir <- gsub("filtered_feature_bc_matrix", "raw_feature_bc_matrix", filt.dir)
  if (!dir.exists(raw.dir)) {
    raw.dir <- file.path("/cluster/projects/aokigroup/Oshima/Thymus_data/raw", sample_id)
  }
  if (!dir.exists(raw.dir) || !dir.exists(filt.dir)) {
    warning(paste("Skipping", sample_id, "- raw or filtered dir not found."))
    next
  }

  filt <- Read10X(filt.dir)
  raw <- Read10X(raw.dir)

  raw <- raw[, colnames(raw) %in% colnames(filt)]
  raw <- raw[rownames(raw) %in% rownames(filt), ]

  sc <- SoupChannel(tod = raw, toc = filt)

  tmp <- CreateSeuratObject(counts = filt, project = sample_id)
  tmp <- NormalizeData(tmp)
  tmp <- FindVariableFeatures(tmp)
  tmp <- ScaleData(tmp)
  tmp <- RunPCA(tmp)
  tmp <- RunUMAP(tmp, dims = 1:20)
  tmp <- FindNeighbors(tmp, dims = 1:20)
  tmp <- FindClusters(tmp, resolution = 0.5)

  umap <- tmp@reductions$umap@cell.embeddings
  clusters <- tmp$seurat_clusters
  names(clusters) <- gsub(paste0("^", sample_id, "_"), "", names(clusters))
  rownames(umap) <- gsub(paste0("^", sample_id, "_"), "", rownames(umap))

  sc <- setClusters(sc, clusters)
  sc <- setDR(sc, umap)
  sc <- autoEstCont(sc)
  cleaned <- adjustCounts(sc)

  cleaned.seurat <- CreateSeuratObject(counts = cleaned, project = sample_id)
  cleaned.seurat <- RenameCells(cleaned.seurat, add.cell.id = sample_id)

  sc.obj <- if (is.null(sc.obj)) cleaned.seurat else merge(sc.obj, cleaned.seurat)

  DropletUtils:::write10xCounts(paste0(out.dir, "/", sample_id), cleaned)
}
```


# 3. Add Metadata (Full)
```{r}
lib.info <- read_xlsx('metadata/lib_ALCL_Thymus_RLN.xlsx')
sc.obj$Sample <- sc.obj$orig.ident
meta <- sc.obj@meta.data
meta$CellBarcode <- rownames(meta)
meta_merged <- left_join(meta, lib.info, by = "Sample")
rownames(meta_merged) <- meta_merged$CellBarcode
meta_merged$CellBarcode <- NULL
sc.obj@meta.data <- meta_merged
```


# 4. QC Filtering (mt% + MAD)
```{r}
sc.obj[["percent.mt"]] <- PercentageFeatureSet(sc.obj, pattern = "^MT-")
sc.obj[["percent.rb"]] <- PercentageFeatureSet(sc.obj, pattern = "^RP[LS]")

keep.nGene <- rep(NA, ncol(sc.obj))
for (s in unique(sc.obj$orig.ident)) {
  s.idx <- which(sc.obj$orig.ident == s)
  s.median <- median(sc.obj$nFeature_RNA[s.idx])
  s.mad <- mad(sc.obj$nFeature_RNA[s.idx])
  keep.nGene[s.idx] <- (sc.obj$nFeature_RNA[s.idx] >= s.median - 4*s.mad) &
                       (sc.obj$nFeature_RNA[s.idx] <= s.median + 4*s.mad)
}
sc.obj$keep_nGene <- keep.nGene

sc.obj <- subset(sc.obj, subset = percent.mt < 20 & keep_nGene)

cell_counts <- table(sc.obj$orig.ident)
samples_to_keep <- names(cell_counts[cell_counts >= 1000])
sc.obj <- subset(sc.obj, orig.ident %in% samples_to_keep)

saveRDS(sc.obj, "data/250625_merged_scobj_qc_filtered_applied.rds")
write.csv(as.data.frame(table(sc.obj$orig.ident)), "data/cell_counts_after_QC_applied.csv")
```


# 5. Normalize, Variable Features, PCA, UMAP, Clustering
```{r}
sc.obj <- NormalizeData(sc.obj)
sc.obj <- FindVariableFeatures(sc.obj, selection.method = "vst", nfeatures = 2000)
top10 <- head(VariableFeatures(sc.obj), 10)

png("figures/250625_top10_variable_genes.png", width = 800, height = 600)
plot <- VariableFeaturePlot(sc.obj)
plot <- LabelPoints(plot, points = top10, repel = TRUE, size = 5) +
  theme(axis.text = element_text(size = 20),
        axis.title = element_text(size = 24),
        legend.title = element_text(size = 24),
        legend.text = element_text(size = 20)) +
  ggtitle('Top 10 Variable Genes')
print(plot)
dev.off()

hvf.info <- HVFInfo(sc.obj)
hvf.info$gene <- rownames(hvf.info)
hvf.ranked <- hvf.info[order(-hvf.info$variance.standardized), ]
write.csv(hvf.ranked, "data/all_variable_features_ranked.csv", row.names = FALSE)

sc.obj <- ScaleData(sc.obj, verbose = FALSE)
sc.obj <- RunPCA(sc.obj, npcs = 30, verbose = FALSE)
sc.obj <- RunUMAP(sc.obj, reduction = "pca", dims = 1:30, reduction.name = 'umap_GEX')
sc.obj <- FindNeighbors(sc.obj, reduction = "pca", dims = 1:30)
sc.obj <- FindClusters(sc.obj, resolution = 0.8)
```


# 6. UMAP Plots (Clusters, Metadata)
```{r}
png("figures/250626_umap_Seurat_clusters_before_correction.png", width = 800, height = 600)
DimPlot(sc.obj, group.by = 'seurat_clusters', reduction = 'umap_GEX', label = TRUE, label.size = 10, raster=FALSE)
dev.off()

png("figures/250626_umap_StudyID_before_correction.png", width = 1200, height = 900)
DimPlot(sc.obj, group.by = 'StudyID', reduction = 'umap_GEX', label = FALSE, label.size = 10, raster=FALSE) +
  theme(legend.text = element_text(size = 30),
        axis.title = element_text(size = 30),
        axis.text = element_text(size = 30))
dev.off()

png("figures/250626_umap_Category1_before_correction.png", width = 1200, height = 900)
DimPlot(sc.obj, group.by = 'Category1', reduction = 'umap_GEX', label = TRUE, label.size = 10, raster=FALSE) +
  theme(legend.text = element_text(size = 30),
        axis.title = element_text(size = 30),
        axis.text = element_text(size = 30))
dev.off()

png("figures/250626_umap_Category2_before_correction.png", width = 1200, height = 900)
DimPlot(sc.obj, group.by = 'Category2', reduction = 'umap_GEX', label = FALSE, label.size = 10, raster=FALSE) +
  theme(legend.text = element_text(size = 30),
        axis.title = element_text(size = 30),
        axis.text = element_text(size = 30))
dev.off()
```


# 7. Save Final Clustered Seurat Object
```{r}
dir.create("data/Seurat_objects", recursive = TRUE, showWarnings = FALSE)
save(sc.obj, file = "data/Seurat_objects/sc.obj_clustered.Rdata")
```

# TO-DO: 
## Make a clear list of different steps and libraries used in HL Rmd and ALSF Rmd
## Use Raw data from RawDir, not filtered_feature_bc_matrix
## Use SoupX to decontaminate the raw data, then create a Seurat object
## Explore how to use renv for reproducible environments
## Do Batch Correction with Harmony, compare with previous clustering, and save the corrected Seurat object
## Manual cell type annotation: mac_marker.txt, component_genes, nkgdt files.

